bitswap/client/internal/session/session.go

```diff
@@ -78,6 +81,41 @@ type ProviderFinder interface {
 	FindProvidersAsync(ctx context.Context, k cid.Cid) <-chan peer.ID
 }
 
+type FindAllProviders struct {
+	Router bsnet.BitSwapNetwork
+}
+
+func (r FindAllProviders) FindProvidersAsync(ctx context.Context, k cid.Cid) <-chan peer.ID {
+	ctx, span := internal.StartSpan(ctx, "FindAllProviders.FindProvidersAsync", trace.WithAttributes(attribute.Stringer("cid", k)))
+	providers := r.Router.FindProvidersAsync(ctx, k, 0)
+	provCh := make(chan peer.ID)
+	wg := &sync.WaitGroup{}
+	for p := range providers {
+		wg.Add(1)
+		go func(p peer.ID) {
+			defer wg.Done()
+			span.AddEvent("FoundProvider", trace.WithAttributes(attribute.Stringer("peer", p)))
+			err := r.Router.ConnectTo(ctx, p)
+			if err != nil {
+				span.RecordError(err, trace.WithAttributes(attribute.Stringer("peer", p)))
+				log.Debugf("failed to connect to provider %s: %s", p, err)
+				return
+			}
+			span.AddEvent("ConnectedToProvider", trace.WithAttributes(attribute.Stringer("peer", p)))
+			select {
+			case provCh <- p:
+			case <-ctx.Done():
+				return
+			}
+		}(p)
+	}
+	go func() {
+		wg.Wait()
+		close(provCh)
+	}()
+	return provCh
```

Comment:
Note: this was copied out of the ProviderQueryManager because the internals (or at least the tests) seem to rely on having the connections made explicit here. It seems reasonable to extract this up a level out of the routing component though.